src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -84,9 +85,10 @@ private function completeDefinition($id, Definition $definition)
                 if (null === $this->types) {
                     $this->populateAvailableTypes();
                 }
-
                 if (isset($this->types[$typeHint->name])) {
                     $value = new Reference($this->types[$typeHint->name]);
+                } elseif (isset($this->serviceTypes[$typeHint->name]) && isset($this->notGuessableTypes[$typeHint->name])) {
```

Comment:
I slightly simplified the code now, if the type is in notGuessableTypes we should throw an exception.